File: web/src/Web.App/Domain/Benchmark/CustomData.cs
using System.Diagnostics.CodeAnalysis;
using Web.App.ViewModels;
namespace Web.App.Domain;

[ExcludeFromCodeCoverage]
public record CustomData
{
    // Administrative supplies
    public decimal? AdministrativeSuppliesCosts { get; set; }

    // Catering
    public decimal? CateringStaffCosts { get; set; }
    public decimal? CateringSuppliesCosts { get; set; }

    // Educational supplies
    public decimal? ExaminationFeesCosts { get; set; }
    public decimal? LearningResourcesNonIctCosts { get; set; }

    // IT
    public decimal? LearningResourcesIctCosts { get; set; }

    // Non-educational support staff
    public decimal? AdministrativeClericalStaffCosts { get; set; }
    public decimal? AuditorsCosts { get; set; }
    public decimal? OtherStaffCosts { get; set; }
    public decimal? ProfessionalServicesNonCurriculumCosts { get; set; }

    // Premises and services
    public decimal? CleaningCaretakingCosts { get; set; }
    public decimal? MaintenancePremisesCosts { get; set; }
    public decimal? OtherOccupationCosts { get; set; }
    public decimal? PremisesStaffCosts { get; set; }

    // Teaching and teaching support
    public decimal? AgencySupplyTeachingStaffCosts { get; set; }
    public decimal? EducationSupportStaffCosts { get; set; }
    public decimal? EducationalConsultancyCosts { get; set; }
    public decimal? SupplyTeachingStaffCosts { get; set; }
    public decimal? TeachingStaffCosts { get; set; }

    // Utilities
    public decimal? EnergyCosts { get; set; }
    public decimal? WaterSewerageCosts { get; set; }

    // Other costs
    public decimal? DirectRevenueFinancingCosts { get; set; }
    public decimal? GroundsMaintenanceCosts { get; set; }
    public decimal? IndirectEmployeeExpenses { get; set; }
    public decimal? InterestChargesLoanBank { get; set; }
    public decimal? OtherInsurancePremiumsCosts { get; set; }
    public decimal? PrivateFinanceInitiativeCharges { get; set; }
    public decimal? RentRatesCosts { get; set; }
    public decimal? SpecialFacilitiesCosts { get; set; }
    public decimal? StaffDevelopmentTrainingCosts { get; set; }
    public decimal? StaffRelatedInsuranceCosts { get; set; }
    public decimal? SupplyTeacherInsurableCosts { get; set; }

    // Totals
    public decimal? TotalIncome { get; set; }
    public decimal? TotalExpenditure { get; set; }
    public decimal? RevenueReserve { get; set; }

    // Non-financial data
    public decimal? NumberOfPupilsFte { get; set; }
    public decimal? FreeSchoolMealPercent { get; set; }
    public decimal? SpecialEducationalNeedsPercent { get; set; }
    public decimal? FloorArea { get; set; }

    // Workforce data
    public decimal? WorkforceFte { get; set; }
    public decimal? TeachersFte { get; set; }
    public decimal? QualifiedTeacherPercent { get; set; }
    public decimal? SeniorLeadershipFte { get; set; }
    public decimal? TeachingAssistantsFte { get; set; }
    public decimal? NonClassroomSupportStaffFte { get; set; }
    public decimal? AuxiliaryStaffFte { get; set; }
    public decimal? WorkforceHeadcount { get; set; }

    public void Merge(ICustomDataViewModel viewModel)
    {
        switch (viewModel)
        {
            case IFinancialDataCustomDataViewModel financialViewModel:
                Merge(financialViewModel);
                break;
            case INonFinancialDataCustomDataViewModel nonFinancialViewModel:
                Merge(nonFinancialViewModel);
                break;
            case IWorkforceDataCustomDataViewModel workforceViewModel:
                Merge(workforceViewModel);
                break;
        }
    }

    private void Merge(IFinancialDataCustomDataViewModel viewModel)
    {
        // Administrative supplies
        AdministrativeSuppliesCosts = viewModel.AdministrativeSuppliesCosts;

        // Catering
        CateringStaffCosts = viewModel.CateringStaffCosts;
        CateringSuppliesCosts = viewModel.CateringSuppliesCosts;

        // Educational supplies
        ExaminationFeesCosts = viewModel.ExaminationFeesCosts;
        LearningResourcesNonIctCosts = viewModel.LearningResourcesNonIctCosts;

        // IT
        LearningResourcesIctCosts = viewModel.LearningResourcesIctCosts;

        // Non-educational support staff
        AdministrativeClericalStaffCosts = viewModel.AdministrativeClericalStaffCosts;
        AuditorsCosts = viewModel.AuditorsCosts;
        OtherStaffCosts = viewModel.OtherStaffCosts;
        ProfessionalServicesNonCurriculumCosts = viewModel.ProfessionalServicesNonCurriculumCosts;

        // Premises and services
        CleaningCaretakingCosts = viewModel.CleaningCaretakingCosts;
        MaintenancePremisesCosts = viewModel.MaintenancePremisesCosts;
        OtherOccupationCosts = viewModel.OtherOccupationCosts;
        PremisesStaffCosts = viewModel.PremisesStaffCosts;

        // Teaching and teaching support
        AgencySupplyTeachingStaffCosts = viewModel.AgencySupplyTeachingStaffCosts;
        EducationSupportStaffCosts = viewModel.EducationSupportStaffCosts;
        EducationalConsultancyCosts = viewModel.EducationalConsultancyCosts;
        SupplyTeachingStaffCosts = viewModel.SupplyTeachingStaffCosts;
        TeachingStaffCosts = viewModel.TeachingStaffCosts;

        // Utilities
        EnergyCosts = viewModel.EnergyCosts;
        WaterSewerageCosts = viewModel.WaterSewerageCosts;

        // Other costs
        DirectRevenueFinancingCosts = viewModel.DirectRevenueFinancingCosts;
        GroundsMaintenanceCosts = viewModel.GroundsMaintenanceCosts;
        IndirectEmployeeExpenses = viewModel.IndirectEmployeeExpenses;
        InterestChargesLoanBank = viewModel.InterestChargesLoanBank;
        OtherInsurancePremiumsCosts = viewModel.OtherInsurancePremiumsCosts;
        PrivateFinanceInitiativeCharges = viewModel.PrivateFinanceInitiativeCharges;
        RentRatesCosts = viewModel.RentRatesCosts;
        SpecialFacilitiesCosts = viewModel.SpecialFacilitiesCosts;
        StaffDevelopmentTrainingCosts = viewModel.StaffDevelopmentTrainingCosts;
        StaffRelatedInsuranceCosts = viewModel.StaffRelatedInsuranceCosts;
        SupplyTeacherInsurableCosts = viewModel.SupplyTeacherInsurableCosts;

        // Totals
        TotalIncome = viewModel.TotalIncome;
        TotalExpenditure = viewModel.TotalExpenditure;
        RevenueReserve = viewModel.RevenueReserve;
    }

    private void Merge(INonFinancialDataCustomDataViewModel viewModel)
    {
        NumberOfPupilsFte = viewModel.NumberOfPupilsFte;
        FreeSchoolMealPercent = viewModel.FreeSchoolMealPercent;
        SpecialEducationalNeedsPercent = viewModel.SpecialEducationalNeedsPercent;
        FloorArea = viewModel.FloorArea;
    }

    private void Merge(IWorkforceDataCustomDataViewModel viewModel)
    {
        WorkforceFte = viewModel.WorkforceFte;
        TeachersFte = viewModel.TeachersFte;
        QualifiedTeacherPercent = viewModel.QualifiedTeacherPercent;
        SeniorLeadershipFte = viewModel.SeniorLeadershipFte;
        TeachingAssistantsFte = viewModel.TeachingAssistantsFte;
        NonClassroomSupportStaffFte = viewModel.NonClassroomSupportStaffFte;
        AuxiliaryStaffFte = viewModel.AuxiliaryStaffFte;
        WorkforceHeadcount = viewModel.WorkforceHeadcount;
    }
}

File: data-pipeline/src/pipeline/pre_processing/ancillary/custom_data.py
import pandas as pd

import pipeline.config as config
import pipeline.pre_processing.common.mappings as mappings


def update_custom_data(
    existing_data: pd.DataFrame,
    custom_data: dict,
    target_urn: int,
) -> pd.DataFrame:
    """
    Update existing financial data with custom data.

    This will overwrite financial information for a specific row with
    data provided; additionally, _all_ "central services" information
    (less the "Financial Position" as this is non-numeric) will be set
    to zero, again for that row only.

    It is assumed that the updated custom-data constitutes a full year
    of financial information and as a result, `Partial Years Present`
    will _always_ be set to _false_ for the row in question.

    Note: only a subset of the custom fields may be present in the
    inbound message; only a subset of mapped columns may be present in
    the existing data. Equally, the data will only be updated if the
    target is present in the existing data.

    :param existing_data: existing, pre-processed data
    :param custom_data: custom financial information
    :param target_urn: specific row to update
    :return: updated data
    """

    if target_urn not in existing_data.index:
        return existing_data

    custom_to_columns = {
        "administrativeSuppliesNonEducationalCosts": "Administrative supplies_Administrative supplies (non educational)",
        "cateringStaffCosts": "Catering staff and supplies_Catering staff",
        "cateringSuppliesCosts": "Catering staff and supplies_Catering supplies",
        "incomeCateringServices": "Income_Catering services",
        "examinationFeesCosts": "Educational supplies_Examination fees",
        "learningResourcesNonIctCosts": "Educational supplies_Learning resources (not ICT equipment)",
        "learningResourcesIctCosts": "Educational ICT_ICT learning resources",
        "administrativeClericalStaffCosts": "Non-educational support staff and services_Administrative and clerical staff",
        "auditorsCosts": "Non-educational support staff and services_Audit cost",
        "otherStaffCosts": "Non-educational support staff and services_Other staff",
        "professionalServicesNonCurriculumCosts": "Non-educational support staff and services_Professional services (non-curriculum)",
        "cleaningCaretakingCosts": "Premises staff and services_Cleaning and caretaking",
        "maintenancePremisesCosts": "Premises staff and services_Maintenance of premises",
        "otherOccupationCosts": "Premises staff and services_Other occupation costs",
        "premisesStaffCosts": "Premises staff and services_Premises staff",
        "agencySupplyTeachingStaffCosts": "Teaching and Teaching support staff_Agency supply teaching staff",
        "educationSupportStaffCosts": "Teaching and Teaching support staff_Education support staff",
        "educationalConsultancyCosts": "Teaching and Teaching support staff_Educational consultancy",
        "supplyTeachingStaffCosts": "Teaching and Teaching support staff_Supply teaching staff",
        "teachingStaffCosts": "Teaching and Teaching support staff_Teaching staff",
        "energyCosts": "Utilities_Energy",
        "waterSewerageCosts": "Utilities_Water and sewerage",
        "directRevenueFinancingCosts": "Other costs_Direct revenue financing",
        "groundsMaintenanceCosts": "Other costs_Grounds maintenance",
        "indirectEmployeeExpenses": "Other costs_Indirect employee expenses",
        "interestChargesLoanBank": "Other costs_Interest charges for loan and bank",
        "otherInsurancePremiumsCosts": "Other costs_Other insurance premiums",
        "privateFinanceInitiativeCharges": "Other costs_PFI charges",
        "rentRatesCosts": "Other costs_Rent and rates",
        "specialFacilitiesCosts": "Other costs_Special facilities",
        "staffDevelopmentTrainingCosts": "Other costs_Staff development and training",
        "staffRelatedInsuranceCosts": "Other costs_Staff-related insurance",
        "supplyTeacherInsurableCosts": "Other costs_Supply teacher insurance",
        "totalIncome": "Total Income",
        "totalExpenditure": "Total Expenditure",
        "revenueReserve": "Revenue reserve",
        "totalPupils": "Number of pupils",
        "percentFreeSchoolMeals": "Percentage Free school meals",
        "percentSpecialEducationNeeds": "Percentage SEN",
        "totalInternalFloorArea": "Total Internal Floor Area",
        "workforceFTE": "Total School Workforce (Full-Time Equivalent)",
        "teachersFTE": "Total Number of Teachers (Full-Time Equivalent)",
        "percentTeacherWithQualifiedStatus": "Teachers with Qualified Teacher Status (%) (Headcount)",
        "seniorLeadershipFTE": "Total Number of Teachers in the Leadership Group (Full-time Equivalent)",
        "teachingAssistantFTE": "Total Number of Teaching Assistants (Full-Time Equivalent)",
        "nonClassroomSupportStaffFTE": "NonClassroomSupportStaffFTE",
        "auxiliaryStaffFTE": "Total Number of Auxiliary Staff (Full-Time Equivalent)",
        "workforceHeadcount": "Total School Workforce (Headcount)",
    }

    custom_present = [
        custom
        for custom in custom_to_columns.keys()
        if custom in custom_data and custom_to_columns[custom] in existing_data.columns
    ]

    existing_columns = [custom_to_columns[custom] for custom in custom_present]
    custom_values = [custom_data[custom] for custom in custom_present]
    existing_data.loc[target_urn, existing_columns] = custom_values

    central_services_columns = [
        column
        for column in existing_data.columns
        if column.endswith("_CS") and column != "Financial Position_CS"
    ]
    central_services_values = [0.0] * len(central_services_columns)
    existing_data.loc[
        target_urn,
        central_services_columns,
    ] = central_services_values

    existing_data.loc[target_urn] = _post_process_custom(
        target_data=existing_data.loc[[target_urn]]
    ).loc[target_urn]

    existing_data.loc[target_urn, "Partial Years Present"] = False

    return existing_data


# TODO: this could be replaced by `pipeline.pre_processing.total_per_unit`.
def _post_process_custom(
    target_data: pd.DataFrame,
) -> pd.DataFrame:
    """
    Recalculate total and per-unit costs for custom data.

    Total and per-unit costs are set to zero before recalculating.

    Note: largely repeats some of the logic from
    :func:`build_maintained_school_data`; for this reason it accepts
    a :class:`DataFrame` but must only include the URN of interest.

    :param existing_data: existing, pre-processed data
    :return: updated data
    """
    zero_columns = [
        column
        for column in target_data.columns
        for category, _ in config.rag_category_settings.items()
        if column.startswith(category) and column.endswith(("_Per Unit", "_Total"))
    ]
    zero_column_indices = [target_data.columns.get_loc(c) for c in zero_columns]
    zero_column_values = [0.0] * len(zero_column_indices)
    target_data.iloc[0, zero_column_indices] = zero_column_values

    # TODO: `_Net Costs` need to be recalculated as per line 1152.
    catering_net_costs = target_data["Catering staff and supplies_Net Costs"].copy()
    target_data["Catering staff and supplies_Net Costs"] = 0.0

    for category, settings in config.rag_category_settings.items():
        basis_data = target_data[
            (
                "Number of pupils"
                if settings["type"] == "Pupil"
                else "Total Internal Floor Area"
            )
        ]
        target_data = mappings.map_cost_series(category, target_data, basis_data)

    target_data["Catering staff and supplies_Net Costs"] = catering_net_costs

    return target_data
